subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptClassPathResolver.java

```diff
@@ -64,29 +94,55 @@ public void prepareClassPath(Configuration configuration, DependencyHandler depe
     }
 
-    @Override
-    public ClassPath resolveClassPath(Configuration classpathConfiguration) {
-        return classpathTransformer.transform(
-            resolveToClasspathWithoutGradleApi(classpathConfiguration),
-            CachedClasspathTransformer.StandardTransform.BuildLogic
+    private void registerTransform(DependencyHandler dependencyHandler, Class<? extends BaseInstrumentingArtifactTransform> transform, String instrumentedAttribute) {
+        dependencyHandler.registerTransform(
+            transform,
+            spec -> {
+                spec.getFrom().attribute(INSTRUMENTED_ATTRIBUTE, NOT_INSTRUMENTED_ATTRIBUTE);
+                spec.getTo().attribute(INSTRUMENTED_ATTRIBUTE, instrumentedAttribute);
+                spec.parameters(parameters -> {
+                    parameters.getAgentSupported().set(agentStatus.isAgentInstrumentationEnabled());
+                    parameters.getMaxSupportedJavaVersion().set(AsmConstants.MAX_SUPPORTED_JAVA_VERSION);
+                    parameters.getUpgradedPropertiesHash().set(gradleCoreInstrumentingTypeRegistry.getUpgradedPropertiesHash().map(Object::toString).orElse(null));
+                });
+            }
         );
     }
 
-    private static ClassPath resolveToClasspathWithoutGradleApi(Configuration configuration) {
-        ArtifactView view = configuration.getIncoming().artifactView(config -> {
-            config.componentFilter(DefaultScriptClassPathResolver::removeGradleApi);
-        });
-        return DefaultClassPath.of(view.getFiles());
+    @Override
+    public ClassPath resolveClassPath(Configuration classpathConfiguration, DependencyHandler dependencyHandler, ConfigurationContainer configContainer) {
+        FileCollection instrumentedExternalDependencies = getInstrumentedExternalDependencies(classpathConfiguration);
+        FileCollection instrumentedProjectDependencies = getInstrumentedProjectDependencies(classpathConfiguration);
+        return TransformedClassPath.handleInstrumentingArtifactTransform(DefaultClassPath.of(instrumentedExternalDependencies.plus(instrumentedProjectDependencies)));
+    }
+
+    private static FileCollection getInstrumentedExternalDependencies(Configuration classpathConfiguration) {
+        return classpathConfiguration.getIncoming().artifactView((Action<? super ArtifactView.ViewConfiguration>) config -> {
+            config.attributes(it -> it.attribute(INSTRUMENTED_ATTRIBUTE, INSTRUMENTED_EXTERNAL_DEPENDENCY_ATTRIBUTE));
+            config.componentFilter(componentId -> {
+                if (componentId instanceof OpaqueComponentIdentifier) {
+                    DependencyFactoryInternal.ClassPathNotation classPathNotation = ((OpaqueComponentIdentifier) componentId).getClassPathNotation();
+                    return !DefaultScriptClassPathResolver.NO_GRADLE_API.contains(classPathNotation);
+                }
```

Comment:
:nail_care: maybe it is still worth it to have a separate "isGradleApi" method.

Reply:
Makes sense, done